### PR TITLE
Publish domain events as business events to AppInsights

### PR DESCRIPTION
**Pull request by sldblog**
## What does this pull request do?

For all events that go through `SNSPublisher`, also send them to AppInsights as business events.

I considered pulling the IDs out of the `additionalInformation` field but didn't want to add too much extra complexity

image

## What is the intent behind these changes?

Domain events are our pivotal events, so it would be good to be able to explicitly count/aggregate them without resolving them from other indicators

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt [hunk 1]
```diff
@@ -1,25 +1,38 @@
...
   private val client: SnsClient,
   private val objectMapper: ObjectMapper,
+  private val telemetryClient: TelemetryClient,
   @Value("\${aws.sns.enabled}") private val enabled: Boolean,
   @Value("\${aws.sns.topic.arn}") private val arn: String,
 ) {
   companion object : KLogging()
 
-  fun publish(event: EventDTO) {
+  fun publish(referralId: UUID, actor: AuthUser?, event: EventDTO) {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt:24]: nit: non nullable actor would be nice

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt [hunk 2]
```diff
@@ -1,25 +1,38 @@
...
     if (enabled) {
       buildRequestAndPublish(event)
+      telemetryClient.trackEvent(
+        "InterventionsDomainEvent",
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt:28]: out of interest - using this type of naming scheme, what other event types could you envisage?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt [hunk 3]
```diff
@@ -1,25 +1,38 @@
...
   private val objectMapper: ObjectMapper,
+  private val telemetryClient: TelemetryClient,
   @Value("\${aws.sns.enabled}") private val enabled: Boolean,
   @Value("\${aws.sns.topic.arn}") private val arn: String,
 ) {
   companion object : KLogging()
 
-  fun publish(event: EventDTO) {
+  fun publish(referralId: UUID, actor: AuthUser?, event: EventDTO) {
     if (enabled) {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt:25]: nit: AI event should probably trigger even if SNS is disabled?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SNSService.kt [hunk 4]
```diff
@@ -91,7 +91,8 @@ class SNSAppointmentService(
           event.appointment.attendanceSubmittedAt!!,
           mapOf("serviceUserCRN" to referral.serviceUserCRN, "referralId" to referral.id)
         )
-        snsPublisher.publish(snsEvent)
+        // FIXME we don't know who submits these -- needs explicit actor
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/SNSService.kt:94]: agree - we do not have a field for `attendanceSubmittedby` - we should add one

**Review by tomsmyers — APPROVED** [review]
nice !

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt (hunk 1 above)
Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt:24]: yeah, we can do that as soon as we add the `appointmentAttendedRecordedBy` and similar "by" fields on appointment

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt (hunk 3 above)
Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt:25]: hmm. yes

Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt:25]: will do that in a follow-up PR

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt (hunk 2 above)
Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt:28]: One useful telemetry would be `InterventionsAuthorizationError` I think

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt (hunk 3 above)
Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/component/SNSPublisher.kt:25]: done in #322

**Pull request merged**